feat(traverse): introduce `MaybeBoundIdentifier`

overlookmotel opened the pull request:
`MaybeBoundIdentifier` is similar to `BoundIdentifier`, but can be used where the identifier may or may not have have a `SymbolId` (may or may not be bound).

Typical usage:

```rs
// Create `MaybeBoundIdentifier` from an existing `IdentifierReference`
let binding = MaybeBoundIdentifier::from_identifier_reference(ident, ctx);

// Generate `IdentifierReference`s and insert them into AST
assign_expr.left = binding.create_write_target(ctx);
assign_expr.right = binding.create_read_expression(ctx);
```

overlookmotel commented:
* **#7266** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7266?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7265** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7265?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented:
@Boshen This API is intended for your spread transform use case.

Boshen commented:
Confirmed working as intended.

```rust
        let maybe_bound_identifier = if let Expression::Identifier(ident) = init {
            MaybeBoundIdentifier::from_identifier_reference(ident, ctx)
        } else {
            let bound_identifier =
                ctx.generate_uid_in_current_scope_based_on_node(init, builder.symbol_flags);
            let id = bound_identifier.create_binding_pattern(ctx);
            let init = ctx.ast.move_expression(init);
            builder.decls.push(ctx.ast.variable_declarator(SPAN, decl.kind, id, Some(init), false));
            bound_identifier.to_maybe_bound_identifier()
        };
```

overlookmotel committed:
feat(traverse): introduce `MaybeBoundIdentifier` (#7265)

`MaybeBoundIdentifier` is similar to `BoundIdentifier`, but can be used where the identifier may or may not have have a `SymbolId` (may or may not be bound).

Typical usage:

```rs
// Create `MaybeBoundIdentifier` from an existing `IdentifierReference`
let binding = MaybeBoundIdentifier::from_identifier_reference(ident, ctx);

// Generate `IdentifierReference`s and insert them into AST
assign_expr.left = binding.create_write_target(ctx);
assign_expr.right = binding.create_read_expression(ctx);
```
